File: tests/python/test_qqe_weighted_oscillator.py
import numpy as np
import pytest

from rust_comparison import compare_with_rust
from test_utils import load_test_data

try:
    import vector_ta as ta
except ImportError:
    try:
        import my_project as ta
    except ImportError:
        pytest.skip(
            "Python module not built. Run 'maturin develop --features python' first",
            allow_module_level=True,
        )


def manual_qqe_weighted_oscillator(data, length, factor, smooth, weight):
    data = np.asarray(data, dtype=np.float64)
    n = data.shape[0]
    rsi = np.full(n, np.nan, dtype=np.float64)
    trailing_stop = np.full(n, np.nan, dtype=np.float64)

    valid = np.flatnonzero(np.isfinite(data))
    if valid.size == 0:
        return {"rsi": rsi, "trailing_stop": trailing_stop}
    if valid.size < length + 1:
        return {"rsi": rsi, "trailing_stop": trailing_stop}

    alpha = 2.0 / (float(smooth) + 1.0)
    prev_src = data[int(valid[0])]
    prev_rsi = np.nan
    prev_ts = np.nan

    num_count = 0
    den_count = 0
    diff_count = 0
    num_sum = 0.0
    den_sum = 0.0
    diff_sum = 0.0
    num_state = np.nan
    den_state = np.nan
    diff_state = np.nan
    ema_state = np.nan

    for i in range(int(valid[0]) + 1, n):
        current = data[i]
        if not np.isfinite(current):
            prev_src = np.nan
            continue
        if not np.isfinite(prev_src):
            prev_src = current
            continue

        delta = current - prev_src
        w = weight if np.isfinite(prev_rsi) and np.isfinite(prev_ts) and delta * (prev_rsi - prev_ts) > 0.0 else 1.0
        weighted_delta = delta * w
        abs_weighted_delta = abs(weighted_delta)

        num_count += 1
        if num_count < length:
            num_sum += weighted_delta
        elif num_count == length:
            num_sum += weighted_delta
            num_state = num_sum / float(length)
        else:
            num_state = ((num_state * (length - 1.0)) + weighted_delta) / float(length)

        den_count += 1
        if den_count < length:
            den_sum += abs_weighted_delta
        elif den_count == length:
            den_sum += abs_weighted_delta
            den_state = den_sum / float(length)
        else:
            den_state = ((den_state * (length - 1.0)) + abs_weighted_delta) / float(length)

        if np.isfinite(num_state) and np.isfinite(den_state) and den_state != 0.0:
            ratio = num_state / den_state
            ema_state = ratio if not np.isfinite(ema_state) else alpha * ratio + (1.0 - alpha) * ema_state
            curr_rsi = 50.0 * ema_state + 50.0
            rsi[i] = curr_rsi

            if np.isfinite(prev_rsi):
                diff_input = abs(curr_rsi - prev_rsi)
                diff_count += 1
                if diff_count < length:
                    diff_sum += diff_input
                    diff = np.nan
                elif diff_count == length:
                    diff_sum += diff_input
                    diff_state = diff_sum / float(length)
                    diff = diff_state
                else:
                    diff_state = ((diff_state * (length - 1.0)) + diff_input) / float(length)
                    diff = diff_state
            else:
                diff = np.nan

            if np.isfinite(diff):
                crossover = np.isfinite(prev_ts) and curr_rsi > prev_ts and prev_rsi <= prev_ts
                crossunder = np.isfinite(prev_ts) and curr_rsi < prev_ts and prev_rsi >= prev_ts
                if crossover:
                    curr_ts = curr_rsi - diff * factor
                elif crossunder:
                    curr_ts = curr_rsi + diff * factor
                elif np.isfinite(prev_ts):
                    if curr_rsi > prev_ts:
                        curr_ts = max(curr_rsi - diff * factor, prev_ts)
                    else:
                        curr_ts = min(curr_rsi + diff * factor, prev_ts)
                else:
                    curr_ts = curr_rsi
            else:
                curr_ts = curr_rsi

            trailing_stop[i] = curr_ts
            prev_rsi = curr_rsi
            prev_ts = curr_ts

        prev_src = current

    return {"rsi": rsi, "trailing_stop": trailing_stop}


class TestQqeWeightedOscillator:
    @pytest.fixture(scope="class")
    def test_data(self):
        return load_test_data()

    def test_manual_reference_matches_binding(self, test_data):
        close = np.asarray(test_data["close"][:160], dtype=np.float64)
        params = {"length": 14, "factor": 4.236, "smooth": 5, "weight": 2.4}
        result = ta.qqe_weighted_oscillator(close, **params)
        expected = manual_qqe_weighted_oscillator(close, **params)

        np.testing.assert_allclose(
            np.asarray(result["rsi"], dtype=np.float64),
            expected["rsi"],
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(result["trailing_stop"], dtype=np.float64),
            expected["trailing_stop"],
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_matches_rust_reference_generator(self, test_data):
        close = np.asarray(test_data["close"], dtype=np.float64)
        params = {"length": 14, "factor": 4.236, "smooth": 5, "weight": 2.0}
        result = ta.qqe_weighted_oscillator(close, **params)
        compare_with_rust(
            "qqe_weighted_oscillator",
            result,
            "close",
            params,
            rtol=0.0,
            atol=1e-12,
        )

    def test_stream_matches_batch(self, test_data):
        close = np.asarray(test_data["close"][:144], dtype=np.float64)
        params = {"length": 14, "factor": 4.236, "smooth": 5, "weight": 2.0}
        batch = ta.qqe_weighted_oscillator(close, **params)
        stream = ta.QqeWeightedOscillatorStream(**params)

        out_rsi = []
        out_ts = []
        for value in close:
            point = stream.update(float(value))
            if point is None:
                out_rsi.append(np.nan)
                out_ts.append(np.nan)
            else:
                out_rsi.append(point[0])
                out_ts.append(point[1])

        np.testing.assert_allclose(
            np.asarray(out_rsi, dtype=np.float64),
            np.asarray(batch["rsi"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(out_ts, dtype=np.float64),
            np.asarray(batch["trailing_stop"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_batch_metadata_and_first_row_match_single(self, test_data):
        close = np.asarray(test_data["close"][:128], dtype=np.float64)
        result = ta.qqe_weighted_oscillator_batch(
            close,
            length_range=(14, 16, 2),
            factor_range=(4.236, 4.736, 0.5),
            smooth_range=(5, 5, 0),
            weight_range=(2.0, 2.0, 0.0),
        )

        assert result["rows"] == 4
        assert result["cols"] == close.shape[0]
        np.testing.assert_array_equal(result["lengths"], np.array([14, 14, 16, 16], dtype=np.uint64))
        np.testing.assert_allclose(result["factors"], np.array([4.236, 4.736, 4.236, 4.736]))
        np.testing.assert_array_equal(result["smooths"], np.array([5, 5, 5, 5], dtype=np.uint64))
        np.testing.assert_allclose(result["weights"], np.array([2.0, 2.0, 2.0, 2.0]))

        single = ta.qqe_weighted_oscillator(close, length=14, factor=4.236, smooth=5, weight=2.0)
        np.testing.assert_allclose(
            np.asarray(result["rsi"][0], dtype=np.float64),
            np.asarray(single["rsi"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(result["trailing_stop"][0], dtype=np.float64),
            np.asarray(single["trailing_stop"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_invalid_inputs_raise(self, test_data):
        close = np.asarray(test_data["close"][:64], dtype=np.float64)

        with pytest.raises(ValueError, match="invalid length|Invalid length"):
            ta.qqe_weighted_oscillator(close, length=0)
        with pytest.raises(ValueError, match="invalid factor|Invalid factor"):
            ta.qqe_weighted_oscillator(close, factor=-1.0)
        with pytest.raises(ValueError, match="invalid smooth|Invalid smooth"):
            ta.qqe_weighted_oscillator(close, smooth=0)

File: tests/wasm/test_qqe_weighted_oscillator.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { assertArrayClose, loadTestData } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

function manualQqeWeightedOscillator(data, length, factor, smooth, weight) {
    const n = data.length;
    const rsi = new Float64Array(n);
    const trailingStop = new Float64Array(n);
    rsi.fill(NaN);
    trailingStop.fill(NaN);

    let first = -1;
    let validCount = 0;
    for (let i = 0; i < n; i++) {
        if (Number.isFinite(data[i])) {
            if (first === -1) first = i;
            validCount += 1;
        }
    }
    if (first === -1 || validCount < length + 1) {
        return { rsi, trailing_stop: trailingStop };
    }

    const alpha = 2.0 / (smooth + 1.0);
    let prevSrc = data[first];
    let prevRsi = NaN;
    let prevTs = NaN;
    let numCount = 0;
    let denCount = 0;
    let diffCount = 0;
    let numSum = 0.0;
    let denSum = 0.0;
    let diffSum = 0.0;
    let numState = NaN;
    let denState = NaN;
    let diffState = NaN;
    let emaState = NaN;

    for (let i = first + 1; i < n; i++) {
        const current = data[i];
        if (!Number.isFinite(current)) {
            prevSrc = NaN;
            continue;
        }
        if (!Number.isFinite(prevSrc)) {
            prevSrc = current;
            continue;
        }

        const delta = current - prevSrc;
        const w = Number.isFinite(prevRsi) && Number.isFinite(prevTs) && delta * (prevRsi - prevTs) > 0.0 ? weight : 1.0;
        const weightedDelta = delta * w;
        const absWeightedDelta = Math.abs(weightedDelta);

        numCount += 1;
        if (numCount < length) {
            numSum += weightedDelta;
        } else if (numCount === length) {
            numSum += weightedDelta;
            numState = numSum / length;
        } else {
            numState = ((numState * (length - 1.0)) + weightedDelta) / length;
        }

        denCount += 1;
        if (denCount < length) {
            denSum += absWeightedDelta;
        } else if (denCount === length) {
            denSum += absWeightedDelta;
            denState = denSum / length;
        } else {
            denState = ((denState * (length - 1.0)) + absWeightedDelta) / length;
        }

        if (Number.isFinite(numState) && Number.isFinite(denState) && denState !== 0.0) {
            const ratio = numState / denState;
            emaState = Number.isFinite(emaState) ? alpha * ratio + (1.0 - alpha) * emaState : ratio;
            const currRsi = 50.0 * emaState + 50.0;
            rsi[i] = currRsi;

            let diff = NaN;
            if (Number.isFinite(prevRsi)) {
                const diffInput = Math.abs(currRsi - prevRsi);
                diffCount += 1;
                if (diffCount < length) {
                    diffSum += diffInput;
                } else if (diffCount === length) {
                    diffSum += diffInput;
                    diffState = diffSum / length;
                    diff = diffState;
                } else {
                    diffState = ((diffState * (length - 1.0)) + diffInput) / length;
                    diff = diffState;
                }
            }

            let currTs = currRsi;
            if (Number.isFinite(diff)) {
                const crossover = Number.isFinite(prevTs) && currRsi > prevTs && prevRsi <= prevTs;
                const crossunder = Number.isFinite(prevTs) && currRsi < prevTs && prevRsi >= prevTs;
                if (crossover) {
                    currTs = currRsi - diff * factor;
                } else if (crossunder) {
                    currTs = currRsi + diff * factor;
                } else if (Number.isFinite(prevTs)) {
                    currTs = currRsi > prevTs
                        ? Math.max(currRsi - diff * factor, prevTs)
                        : Math.min(currRsi + diff * factor, prevTs);
                }
            }

            trailingStop[i] = currTs;
            prevRsi = currRsi;
            prevTs = currTs;
        }

        prevSrc = current;
    }

    return { rsi, trailing_stop: trailingStop };
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
    testData = loadTestData();
});

test('qqe_weighted_oscillator_js matches manual reference', () => {
    const close = new Float64Array(testData.close.slice(0, 160));
    const result = wasm.qqe_weighted_oscillator_js(close, 14, 4.236, 5, 2.4);
    const expected = manualQqeWeightedOscillator(close, 14, 4.236, 5, 2.4);

    assertArrayClose(result.rsi, expected.rsi, 1e-12, 'rsi mismatch');
    assertArrayClose(result.trailing_stop, expected.trailing_stop, 1e-12, 'trailing stop mismatch');
});

test('qqe_weighted_oscillator_batch first row matches single', () => {
    const close = new Float64Array(testData.close.slice(0, 128));
    const batch = wasm.qqe_weighted_oscillator_batch(close, {
        length_range: [14, 16, 2],
        factor_range: [4.236, 4.736, 0.5],
        smooth_range: [5, 5, 0],
        weight_range: [2.0, 2.0, 0.0],
    });
    const single = wasm.qqe_weighted_oscillator_js(close, 14, 4.236, 5, 2.0);

    assert.strictEqual(batch.rows, 4);
    assert.strictEqual(batch.cols, close.length);
    assert.deepStrictEqual(batch.lengths, [14, 14, 16, 16]);
    assert.deepStrictEqual(batch.factors, [4.236, 4.736, 4.236, 4.736]);
    assert.deepStrictEqual(batch.smooths, [5, 5, 5, 5]);
    assert.deepStrictEqual(batch.weights, [2.0, 2.0, 2.0, 2.0]);

    assertArrayClose(batch.rsi.slice(0, close.length), single.rsi, 1e-12, 'batch rsi mismatch');
    assertArrayClose(
        batch.trailing_stop.slice(0, close.length),
        single.trailing_stop,
        1e-12,
        'batch trailing stop mismatch',
    );
});

test('qqe_weighted_oscillator_js rejects invalid params', () => {
    const close = new Float64Array(testData.close.slice(0, 64));
    assert.throws(() => {
        wasm.qqe_weighted_oscillator_js(close, 0, 4.236, 5, 2.0);
    }, /invalid length/i);
    assert.throws(() => {
        wasm.qqe_weighted_oscillator_js(close, 14, -1.0, 5, 2.0);
    }, /invalid factor/i);
    assert.throws(() => {
        wasm.qqe_weighted_oscillator_js(close, 14, 4.236, 0, 2.0);
    }, /invalid smooth/i);
});
